File: Queue/LinkedList/Java/Queue.java
class Queue<T>{

  Node head;
  Node tail;

  public Queue(){
    this.head = null;
    this.tail = null;
  }

  public void enqueue(T data){
    Node newN = new Node(data, null);
    if(head == null){
      head = tail = newN;
    }else{
      tail.next = newN;
      tail = newN;
    }
  }

  public T dequeue(){
    Node ret = head;
    if(head == null){
      tail = null;
      return null;
    }
    head = head.next;
    return ret.data;
  }

  private class Node{

    T data;
    Node next;

    public Node(T data, Node next){
      this.data = data;
      this.next = next;
    }
  }
}

File: Queue/LinkedList/C++/Queue.cpp
#include "Queue.h"
#include <stdlib.h>
#include <iostream>

Node::Node(int data, Node *next){
    this->data = data;
    this->next = next;
}

Queue::Queue(){
  this->head = nullptr;
  this->tail = nullptr;
}

void Queue::enqueue(int data){
  Node *newN= new Node;
  newN->data = data;
  newN->next = nullptr;
  if(this->head == nullptr){
    this->head = this->tail = newN;
  }else{
    this->tail->next = newN;
    this->tail = newN;
  }
}

int Queue::dequeue(){
  if(this->head == nullptr){
    this->tail = nullptr;
    return -1;
  }
  int ret = this->head->data;
  this->head = this->head->next;
  return ret;
}
